koku/api/report/ocp/query_handler.py

```diff
@@ -31,6 +34,51 @@
 LOG = logging.getLogger(__name__)
 
 
+@dataclass
+class CapacitySubsets:
+    key: str
+    level_key: str
+    resolution: str
+    total: Decimal = Decimal(0)
+    resolution_total: defaultdict = field(default_factory=lambda: defaultdict(Decimal))
+    resolution_level_total: defaultdict = field(default_factory=lambda: defaultdict(lambda: defaultdict(Decimal)))
+    by_level: defaultdict = field(default_factory=lambda: defaultdict(Decimal))
+
+    def add(self, cap_value, usage_start, level_value):
+        self.total += cap_value
+        if self.resolution == "daily":
+            if isinstance(usage_start, datetime.date):
+                usage_start = usage_start.isoformat()
+            self.by_level[level_value] += cap_value
+            self.resolution_level_total[usage_start][level_value] += cap_value
+            self.resolution_total[usage_start] += cap_value
+        elif self.resolution == "monthly":
+            month = usage_start.month
+            self.by_level[level_value] += cap_value
+            self.resolution_total[month] += cap_value
+            self.resolution_level_total[month][level_value] += cap_value
+
+
+def _calculate_unused(row):
+    """Calculates the unused portions of the capacity & request."""
+    # Populate unused request and capacity
+    capacity = row.get("capacity", Decimal(0))
+    if capacity == 0:
```

Comment:
Optional: Is it worth guarding against negative values here too? and on line 77 below. I know its not really likely but if we did get a negative value from a custom report that could be prevented here.